core/profiles/src/main/java/software/amazon/awssdk/profiles/internal/ProfileFileRefresher.java

```diff
@@ -0,0 +1,267 @@
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.Clock;
+import java.time.Instant;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.annotations.SdkTestInternalApi;
+import software.amazon.awssdk.profiles.ProfileFile;
+import software.amazon.awssdk.utils.SdkAutoCloseable;
+import software.amazon.awssdk.utils.cache.CachedSupplier;
+import software.amazon.awssdk.utils.cache.RefreshResult;
+
+/**
+ * Class used for caching and reloading ProfileFile objects from a Supplier.
+ */
+@SdkInternalApi
+public final class ProfileFileRefresher implements SdkAutoCloseable {
+
+    private static final ProfileFileRefreshRecord EMPTY_REFRESH_RECORD = ProfileFileRefreshRecord.builder()
+                                                                                                 .refreshTime(Instant.MIN)
+                                                                                                 .build();
+    private final CachedSupplier<ProfileFileRefreshRecord> profileFileCache;
+    private volatile ProfileFileRefreshRecord currentRefreshRecord;
+    private final Supplier<ProfileFile> profileFile;
+    private final Path profileFilePath;
+    private final Function<RuntimeException, ProfileFile> exceptionHandler;
+    private final Consumer<ProfileFile> onProfileFileReload;
+    private final Clock clock;
+
+    private ProfileFileRefresher(Builder builder) {
+        this.exceptionHandler = builder.exceptionHandler;
+        this.clock = builder.clock;
+        this.profileFile = builder.profileFile;
+        this.profileFilePath = builder.profileFilePath;
+        this.onProfileFileReload = builder.onProfileFileReload;
+        this.profileFileCache = CachedSupplier.builder(this::refreshResult)
+                                              .clock(this.clock)
+                                              .build();
+        this.currentRefreshRecord = EMPTY_REFRESH_RECORD;
+    }
+
+    /**
+     * Builder method to construct instance of ProfileFileRefresher.
+     */
+    public static ProfileFileRefresher.Builder builder() {
+        return new ProfileFileRefresher.Builder();
+    }
+
+    /**
+     * Retrieves the cache value or refreshes it if stale.
+     */
+    public ProfileFile refreshIfStale() {
+        ProfileFileRefreshRecord cachedOrRefreshedRecord = profileFileCache.get();
+        ProfileFile cachedOrRefreshedProfileFile = cachedOrRefreshedRecord.profileFile;
+        if (isNewProfileFile(cachedOrRefreshedProfileFile)) {
+            currentRefreshRecord = cachedOrRefreshedRecord;
+        }
+
+        return cachedOrRefreshedProfileFile;
+    }
+
+    @Override
+    public void close() {
+        profileFileCache.close();
```

Comment:
We only need to close the cached supplier when it is async. I think it's fine to exclude this close for now so that we don't have to deal with managing the lifecycle of a profile file supplier.

Reply:
Removed